ComponentKit/TransactionalDataSources/Common/CKTransactionalComponentDataSourceAppliedChanges.mm

```diff
@@ -48,6 +48,53 @@ - (instancetype)initWithUpdatedIndexPaths:(NSSet *)updatedIndexPaths
   return self;
 }
 
+- (NSIndexPath *)newIndexPathForPreviousIndexPath:(NSIndexPath *)indexPath
+{
+  if ([_removedIndexPaths containsObject:indexPath])
+    return nil;
+  
+  __block NSInteger section = indexPath.section;
+  NSInteger item = indexPath.item;
+  if ([_removedSections containsIndex:section])
+    return nil;
+
+  for (NSIndexPath *sourceIndexPath in _movedIndexPaths.keyEnumerator) {
+    if (sourceIndexPath.section == section && sourceIndexPath.item < indexPath.item) {
+      item--;
+    }
+  }
```

Comment:
Let's include comments explaining each step. Makes reading through this a bit easier.